lib/AST/ShouldPrintForParseableInterface.cpp

```diff
@@ -50,25 +51,33 @@ bool swift::contributesToParentTypeStorage(const AbstractStorageDecl *ASD) {
 /// its layout, and so on.
 class FindReferencedNonPublicTypeDecls: public ASTWalker {
   using TypeDecls = SmallPtrSetImpl<TypeDecl *>;
+  SmallVector<ValueDecl *, 8> workList;
 
   /// Finds all nominal types or typealiases referenced in the type being
   /// walked.
   class FindTypeDecls: public TypeWalker {
     TypeDecls &foundDecls;
+    SmallVectorImpl<ValueDecl *> &workList;
   public:
-    FindTypeDecls(TypeDecls &foundDecls): foundDecls(foundDecls) {}
+    FindTypeDecls(TypeDecls &foundDecls, SmallVectorImpl<ValueDecl *> &workList)
+      : foundDecls(foundDecls), workList(workList) {}
 
     void addIfNonPublic(TypeDecl *typeDecl) {
       if (isPublicOrUsableFromInline(typeDecl))
         return;
-      foundDecls.insert(typeDecl);
+      if (foundDecls.insert(typeDecl).second)
+        workList.push_back(typeDecl);
     }
 
     /// Finds all nominal types or typealiases referenced by the
     Action walkToTypePre(Type type) override {
       if (auto nominal =
           dyn_cast<NominalOrBoundGenericNominalType>(type.getPointer()))
         addIfNonPublic(nominal->getDecl());
+      if (auto enumType = dyn_cast<EnumType>(type.getPointer()))
+        addIfNonPublic(enumType->getDecl());
+      if (auto protocol = dyn_cast<ProtocolType>(type.getPointer()))
+        addIfNonPublic(protocol->getDecl());
```

Comment:
These two should be redundant because they're both NominalOrBoundGenericNominalTypes. What's missing is protocol composition types, unless we treat that as having ProtocolTypes inside it?